[Team - 06] [iOS] Airbnb Map 설정

piggyse opened the pull request:
## 작업내용
- 앱 Flow 파악
- Mapkit을 이용한 MapView띄우기
- CutomAnnotationView 사용 map에 가격띄우기
- Callout 설정

![Simulator Screen Recording - iPhone 13 - 2022-05-25 at 16 49 43](https://user-images.githubusercontent.com/80263729/170213294-178418ac-14f8-4b16-8fe4-90cb9df8c54a.gif)



- - -
### Flow 파악
![스크린샷 2022-05-25 오후 4 51 18](https://user-images.githubusercontent.com/80263729/170210111-2528ba17-0ff2-417c-a072-c4c4e35c9952.png)

- MapView띄우기
Flow를 파악해보니 `request보내는 부분`은 설정값을 바탕으로 `Map과 List가 띄워지기 전`이라 생각해 Map부터 구현하면 좋겠다는 생각에 
Map부분부터 구현을 하게 되었습니다.

- - -
### 구현 설계
![스크린샷 2022-05-25 오후 4 51 34](https://user-images.githubusercontent.com/80263729/170210163-5223988c-1bc4-42e3-926e-2d6a553082e8.png)

- MapView위에 ColletectionView가 띄워지는 형태로 구현을 하려 했습니다.
하지만 MapView가 CollectionView의 Delegate를 설정하는 것은 아닌 것같아 Delegate는 VC에서 처리하도록 설계를 했습니다.
향후, VC가 너무 커지면 이부분도 분리를 생각중입니다.

- - -
### 고민 및 해결
![스크린샷 2022-05-25 오후 4 59 27](https://user-images.githubusercontent.com/80263729/170211826-ba4d406c-5ae4-4755-b4d3-3297093af48f.png)

`고민`
같은 Label안에 다른 폰트가 섞여있는 것을 확인 
`해결` 
`NSMutableAttributedString`을 이용해 해결을 시도해보았습니다.

`고민` 
Custom AnnotationView의 Callout이 설정이 안되던 문제
`해결` 
UILabel이 AnnotationView의 모든 영역을 차지해서 터치가 인식이 안된 문제, AnntationView의 Frame을 UIView와 같이 둠으로써 해결 시도.

leejohy-0223 committed:
initial commit

leejohy-0223 committed:
picture add

leejohy-0223 committed:
Update issue templates

leejohy-0223 committed:
Update README.md

leejohy-0223 committed:
[docs]: img 제거

piggyse committed:
Setting Airbnb Project

contents
- git ignore
- add Dependencies (snapkit, alamofire)
- remove storyboard

piggyse committed:
[feature] Add MapViewController [#2]

Contents
- MapViewController 생성
- MKMapView 생성
- Snapkit Dynamic Dependency 제거

piggyse committed:
[feature] Custom MapView 생성 [#2]

Contents
- MapViewController에 들어갈 CustomMapView 클래스 생성

Todo
- MapView 안의 CollectionView 설정
- CollectView Delegate 생성
- CustomCell 생성

piggyse committed:
[feature] Custom Annotation 생성 [#2]

Contents
- Custom AnnotationView 생성
- HouseInfo dummy data로 구현
- Callout 설정

piggyse committed:
[feature] AddressConverter 추가

Contents
- 주어진 coordinate를 받아서 한국식 주소로 변환
- annotaion의 subtitle로 설정

piggyse committed:
[test] Converter Test코드 추가 [#4]

Contents
- PriceConverter Unit Test 추가
- AddressConverter Unit Test 추가

piggyse committed:
[chore] .xcuserstate 파일 캐쉬 제거

piggyse committed:
[style] Constant 파일 style 수정

Contents
- Nested 구조로 변경 EX) Constant.Pin.size

piggyse committed:
[feature] CollectionView & CardCell MockImage로 테스트

Contents
- 임시값(300)과 house image로 Cell layout 체크

piggyse committed:
[feature] 후기 및 별점 Label 설정 [#8]

Contents
- NSMutableAttributedString 이용 각 Text 값 다르게 setting
- NSMutableAttributedString extension으로Setting함수 추가

piggyse committed:
[feature] 하트버튼 추가 [#8]

Contents
- MapCard 하트버튼 추가

Todo
- 버튼 액션 추가

Piggy-Seob commented on iOS/AirbnbApp/AirbnbApp/Map/MapViewCardCell.swift:
```diff
@@ -0,0 +1,114 @@
+    }
+    
+    // Text별 색상과 속성을 다르게 줌.
+    func setReviewLabel(rating: Double, reviewCount: Int) {
```
이 부분이 한 Label안에 다른 폰트들이 있는것같아 설정해준 부분인데요.. 막상 해놓고 나니 extension으로 함수를 만들었다고 해도 많이 지저분해 보였습니다. 혹시 예시와 같은 View를 이 와같은 방법을 쓰지 않고도 가능하게 할 수 있을까요?

lenaios replied:
하나의 Label로만 구현해야 하는게 아니면 각각 따로 라벨을 생성하고 StackView 등을 활용할 수도 있지 않을까요?

piggyse replied:
제생각에 StackView를 사용하는 방식은 더 많은 View를 만드는 코드를 줘야한다고 생각해서 진행을 안했는데, 다시 한번 장단점을 생각해 보겠습니다!

piggyse committed:
[feature] MapCardCell Label 마무리 [#8]

Contents
- houseNameLabel,pricePerDayLabel 생성
- Layout설정
- 화면 크기에따라 imageView크기 수정

piggyse committed:
[refator] Heartbutton Refactor

Contents
- setImage -> configuartion

lenaios commented on iOS/AirbnbApp/AirbnbApp/Converter/AddressConverter.swift:
```diff
@@ -0,0 +1,40 @@
+            }
+            
+            // 강남대로62길 23
+            if let name: String = placeMark?.name {
+                address += " "
+                address += name
+                }
+            isCompleted(address)
+            
+            }
+        )
```
indent가 안맞는 것 같아요~

lenaios commented on iOS/AirbnbApp/AirbnbApp/Extension/NSMutableAttributedString.swift:
```diff
@@ -0,0 +1,17 @@
+
+import UIKit
+
+extension NSMutableAttributedString {
+    func setting(string: String, fontSize: CGFloat, weight: UIFont.Weight, color: UIColor) -> NSMutableAttributedString {
+        let font = UIFont.systemFont(ofSize: fontSize, weight: weight)
+        let attributes: [NSAttributedString.Key: Any] = [.font: font, .foregroundColor: color]
+        self.append(NSAttributedString(string: string, attributes: attributes))
+        return self
+    }
+}
```
이런 시도 아주 좋네요! 👏

lenaios commented on iOS/AirbnbApp/AirbnbApp/Map/HouseInfo.swift:
```diff
@@ -0,0 +1,14 @@
+
+import CoreLocation
+
+struct HouseInfo {
+    let coordinate: CLLocationCoordinate2D
+    let name: String
+    var address: String = " "
```
address만 변수인 이유는 무엇일까요?

piggyse replied:
coordinate를 바탕으로 `findAddressFromCoordinate` 함수를 실행시켜서 찾기 때문에 초깃값을 가진 변수로 넣어주고 나중에 바꾸어 볼려고 변수로 선언한 것인데 지금 상황에서는 필요없는 코드인것 같습니다!

lenaios commented on iOS/AirbnbApp/AirbnbApp/Map/MapViewCardCell.swift:
```diff
@@ -0,0 +1,160 @@
+        
+    }
+    
+    @objc private func changeImage() {
+        if heartButton.configuration?.image == UIImage(systemName: "heart") {
+            heartButton.configuration?.image = UIImage(systemName: "heart.fill")
+            heartButton.configuration?.baseForegroundColor = .red
+        } else {
+            heartButton.configuration?.image = UIImage(systemName: "heart")
+            heartButton.configuration?.baseForegroundColor = .secondaryLabel
+        }
+    }
```
버튼의 현재 이미지를 기준으로 판단하고 있네요~ heart 유무를 관리하는 객체나 속성이 있어도 괜찮을 것 같습니다. 🙂

lenaios commented on iOS/AirbnbApp/AirbnbApp/SupportingFiles/Constants.swift:
```diff
@@ -0,0 +1,30 @@
+        static let ID: String = "CustomPin"
+    }
+    
+    struct CellID {
+        static let map: String = "Map"
+    }
```
특정 뷰 안에서만 cell register를 위해 사용되는 값이라면, 관련 클래스 안에 선언하는 것도 괜찮아 보입니다 😊

lenaios reviewed (approved):
안녕하세요! @Piggy-Seob 
리뷰에 앞서 모든 코멘트는 반드시 반영하지 않아도 괜찮습니다. :)

- 전체적으로 코드를 깔끔하게 작성하셨네요 👏
- 아직 구현 중이라 코드 중간중간 더미 값들이 보이는데, 이후 이 부분들이 어떻게 수정될지 궁금하네요~
- 의문이 드는 점이나 의견드리고 싶은 부분에 리뷰를 남겼으니 확인 부탁드립니다 😊

직접 merge 해주세요! 수고 많으셨습니다~
